File: 109.convert-sorted-list-to-binary-search-tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST2(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if not nums:
            return None
        middle = len(nums) // 2
        root = TreeNode(nums[middle])
        root.left = self.sortedArrayToBST2(nums[:middle])
        root.right = self.sortedArrayToBST2(nums[middle + 1:])
        return root

    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        nums = []
        while head:
            nums.append(head.val)
            head = head.next
        return self.sortedArrayToBST2(nums)

File: 109.convert-sorted-list-to-binary-search-tree.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST2 = function(nums) {
    if (nums.length === 0) {
        return null;
    }
    var middle = nums.length >> 1;
    var root = new TreeNode(nums[middle]);
    root.left = sortedArrayToBST2(nums.slice(0, middle));
    root.right = sortedArrayToBST2(nums.slice(middle + 1));
    return root;
};

/**
 * @param {ListNode} head
 * @return {TreeNode}
 */
var sortedListToBST = function(head) {
    var nums = [];
    while (head) {
        nums.push(head.val);
        head = head.next;
    }
    return sortedArrayToBST2(nums);
};
